File: Liftr/Pet/PetMarketVisibility.swift
import Foundation

enum PetMarketVisibility {
    static func inventoryQuantity(_ inventory: [PetInventoryRow], itemType: String) -> Int {
        inventory.first(where: { $0.itemType == itemType })?.quantity ?? 0
    }

    static func shouldShowMarketItem(
        _ item: PetMarketItemRow,
        petData: PetFullData?
    ) -> Bool {
        let hasPet = petData?.pet != nil
        let inventory = petData?.inventory ?? []

        switch item.itemType {
        case "pet_egg":
            return !hasPet && inventoryQuantity(inventory, itemType: "pet_egg") == 0
        case "incubator":
            return !hasPet && inventoryQuantity(inventory, itemType: "incubator") == 0
        case "pet_rarity_upgrade":
            guard let pet = petData?.pet,
                  let rarity = PetRarity(databaseValue: pet.rarity),
                  rarity != .mythic else { return false }
            return true
        case "pet_energy_capacity":
            guard let energy = petData?.energy else { return true }
            return energy.max < PetEnergyPricing.maxCapacity
        default:
            if PetFoodItemType.all.contains(item.itemType) {
                return hasPet
            }
            return true
        }
    }

    static func inventoryCategory(for itemType: String) -> String {
        switch itemType {
        case "pet_egg", "incubator":
            return "eggs_devices"
        default:
            return "pet_food"
        }
    }

    static func categoryTitle(_ category: String) -> String {
        switch category {
        case "eggs_devices": return "Eggs & Devices"
        case "pet_food": return "Pet Food"
        case "pet_upgrades": return "Pet Upgrades"
        default: return category.capitalized
        }
    }

    static func hasIncubator(_ inventory: [PetInventoryRow]) -> Bool {
        inventoryQuantity(inventory, itemType: "incubator") > 0
    }

    static func canIncubate(petData: PetFullData?) -> Bool {
        guard petData?.pet == nil else { return false }
        let inventory = petData?.inventory ?? []
        return inventoryQuantity(inventory, itemType: "pet_egg") > 0
            && inventoryQuantity(inventory, itemType: "incubator") > 0
    }
}

File: android/app/src/main/java/com/lilru/liftr/ui/pets/PetMarketVisibility.kt
package com.lilru.liftr.ui.pets

import com.lilru.liftr.data.PetEnergyPricing
import com.lilru.liftr.data.PetFullDataWire
import com.lilru.liftr.data.PetInventoryWire
import com.lilru.liftr.data.PetMarketItemWire

private val FOOD_TYPES = setOf("food_baby", "food_kid", "food_teen", "food_adult", "food_elder")

object PetMarketVisibility {
    fun inventoryQuantity(inventory: List<PetInventoryWire>, itemType: String): Int =
        inventory.firstOrNull { it.itemType == itemType }?.quantity ?: 0

    fun shouldShowMarketItem(item: PetMarketItemWire, petData: PetFullDataWire?): Boolean {
        val hasPet = petData?.pet != null
        val inventory = petData?.inventory.orEmpty()
        return when (item.itemType) {
            "pet_egg" -> !hasPet && inventoryQuantity(inventory, "pet_egg") == 0
            "incubator" -> !hasPet && inventoryQuantity(inventory, "incubator") == 0
            "pet_rarity_upgrade" -> {
                val rarity = petData?.pet?.rarity ?: return false
                PetRarityUpgrade.nextTier(rarity) != null
            }
            "pet_energy_capacity" -> (petData?.energy?.max ?: 5) < PetEnergyPricing.MAX_CAPACITY
            in FOOD_TYPES -> hasPet
            else -> true
        }
    }

    fun inventoryCategory(itemType: String): String = when (itemType) {
        "pet_egg", "incubator" -> "eggs_devices"
        else -> "pet_food"
    }

    fun categoryTitle(category: String): String = when (category) {
        "eggs_devices" -> "Eggs & Devices"
        "pet_food" -> "Pet Food"
        "pet_upgrades" -> "Pet Upgrades"
        else -> category.replaceFirstChar { it.uppercase() }
    }

    fun hasIncubator(inventory: List<PetInventoryWire>): Boolean =
        inventoryQuantity(inventory, "incubator") > 0

    fun canIncubate(petData: PetFullDataWire?): Boolean {
        if (petData?.pet != null) return false
        val inventory = petData?.inventory.orEmpty()
        return inventoryQuantity(inventory, "pet_egg") > 0 &&
            inventoryQuantity(inventory, "incubator") > 0
    }
}
